db/migration/schema_set.go

```diff
@@ -0,0 +1,219 @@
+	schemas stringSet
+	backend schemaSetBackend
+}
+
+func newSchemaSet() *schemaSet {
+	return &schemaSet{
+		schemas: stringSet{},
+		backend: nil, // there is a global schemaSet instance, backend must be lazily initialized
+	}
+}
+
+func (s *schemaSet) init(backend schemaSetBackend) (err error) {
+	log.Notice("post-migration: initializing schema ID set")
+
+	if backend == nil {
+		return errNilBackend
+	}
+
+	s.backend = backend
+
+	schemas, err := backend.loadSchemaIDs()
+	if err != nil {
+		return err
+	}
+
+	s.schemas = schemas
+
+	return nil
+}
+
+func (s *schemaSet) markSchemaID(schemaID string, tx *sql.Tx) (err error) {
+	log.Info("post-migration: marking schema '%s'", schemaID)
+
+	if s.backend == nil {
+		return errNilBackend
+	}
+
+	err = s.backend.persistSchemaID(schemaID, tx)
+	if err != nil {
+		return err
+	}
+
+	s.schemas[schemaID] = true
+	return nil
+}
+
+func (s *schemaSet) getSchemaIDs() []string {
+	schemas := make([]string, 0, len(s.schemas))
+	for schema := range s.schemas {
+		schemas = append(schemas, schema)
+	}
+	return schemas
+}
+
+func (s *schemaSet) removeSchemaID(schemaID string) error {
+	log.Info("post-migration: removing schema '%s'", schemaID)
+
+	if s.backend == nil {
+		return errNilBackend
+	}
+
+	err := s.backend.deleteSchemaID(schemaID)
+	if err != nil {
+		return err
+	}
+
+	delete(s.schemas, schemaID)
+	return nil
+}
+
+type postMigrationEventsBackend struct {
+	db *sql.DB
+}
+
+var _ schemaSetBackend = &postMigrationEventsBackend{}
+
+func newSchemaSetDbBackend(db *sql.DB) *postMigrationEventsBackend {
+	return &postMigrationEventsBackend{db: db}
+}
+
+const schemaSetDbTable = "post_migration_events"
+
+func (b *postMigrationEventsBackend) loadSchemaIDs() (stringSet, error) {
+	log.Debug("post-migration: preloading pending schema IDs")
+
+	tx, err := b.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	schemas, err := b.readSchemaIDsInTx(tx)
+	if err != nil {
+		tx.Rollback()
+		return nil, err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return nil, err
+	}
+
+	return schemas, nil
+}
+
+func (b *postMigrationEventsBackend) readSchemaIDsInTx(tx *sql.Tx) (stringSet, error) {
+	if err := b.ensureTableExists(tx); err != nil {
+		return nil, err
+	}
+
+	const query = "SELECT `schema_id` FROM `" + schemaSetDbTable + "`"
+	rows, err := tx.Query(query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	schemas := make(stringSet)
+
+	for rows.Next() {
+		var schema string
+		if err := rows.Scan(&schema); err != nil {
+			return nil, err
+		}
+
+		schemas[schema] = true
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+
+	log.Debug("post-migration: read %d pending schema IDs", len(schemas))
+	return schemas, nil
+}
+
+func (b *postMigrationEventsBackend) ensureTableExists(tx *sql.Tx) error {
+	log.Debug("post-migration: making sure table `%s` exists", schemaSetDbTable)
+
+	const stmt = "CREATE TABLE IF NOT EXISTS `" + schemaSetDbTable + "` (`schema_id` TEXT NOT NULL)"
```

Comment:
why use such a have type `TEXT`?

Reply:
Actually I didn't even think about it. It shouldn't be a problem because this table should store a couple (2, 3?) record at any time. Anyway I can change it to varchar(255), but I don't know whether we have a limit on schema name length.

Reply:
if this table is small, then it should not be a problem